internal/venice-common/src/main/java/com/linkedin/venice/pubsub/ImmutablePubSubMessage.java

```diff
@@ -13,19 +14,33 @@ public class ImmutablePubSubMessage<K, V> implements PubSubMessage<K, V, Long> {
   private final long timestamp;
   private final int payloadSize;
 
+  private final PubSubMessageHeaders pubSubMessageHeaders;
```

Comment:
I am slightly peeved at having an extra pointer per message, as that will increase the memory footprint of in-flight messages (even though it will almost always be null, null still takes 4B or 8B, depending on whether the heap size makes us use 64 bit pointers or not...).

We could use a primitive boolean instead (which would get padded up to 8 bits, no matter the pointer size). Or we could use extension to have a different class which carries the extra property only when needed.

These seem like premature optimizations so I guess it's fine to leave it as is for now.

Reply:
That is a great point & suggestions. Let me explore this further once I have the 2nd half of this PR ready.